addon/-private/system/relationships/relationships-payloads.js

```diff
@@ -0,0 +1,446 @@
+        inverse: true,
+      }
+    }
+
+  @class RelationshipsPayloads
+*/
+export default class RelationshipsPayloads {
+  constructor(store) {
+    this._store = store;
+    this._map = {};
+  }
+
+  /**
+    Find the payload for the given relationship of the given model.
+
+    Returns a payload entry which will contain the payload for the given
+    relationship and a flag to denote whether this payload was directly pushed
+    or computed from an inverse
+
+    @example
+
+      relationshipsPayloads.get('hobby', 2, 'user') === {
+        {
+          payload: {
+            data: {
+              id: 1,
+              type: 'user'
+            }
+          },
+          // tells us this payload was populated from the inverse side; so if
+          // we're a many side of a relationship we are not fully loaded
+          inverse: true,
+        }
+      }
+
+  */
+  get(modelName, id, relationshipName) {
+    let relationshipPayloads = this._getRelationshipPayloads(modelName, relationshipName);
+    return relationshipPayloads && relationshipPayloads.get(modelName, id, relationshipName);
+  }
+
+  /**
+    Push a model's relationships payload into this cache.
+
+    @example
+
+      let userPayload = {
+        data: {
+          id: 1,
+          type: 'user',
+          relationships: {
+            hobbies: {
+              data: [{
+                id: 2,
+                type: 'hobby'
+              }]
+            }
+          }
+        },
+      };
+      relationshipsPayloads.push('user', 1, userPayload.data.relationships);
+  */
+  push(modelName, id, relationshipsData) {
+    if (!relationshipsData) { return; }
+
+    for (let key in relationshipsData) {
+      let relationshipPayloads = this._getRelationshipPayloads(modelName, key);
+      if (relationshipPayloads) {
+        relationshipPayloads.push(modelName, id, key, relationshipsData[key]);
+      }
+    }
+  }
+
+  unload(modelName, id) {
+    let modelClass = this._store.modelFor(modelName);
+    let relationshipsByName = get(modelClass, 'relationshipsByName');
+    relationshipsByName.forEach((_, relationshipName) => {
+      let relationshipPayloads = this._getRelationshipPayloads(modelName, relationshipName);
+      if (relationshipPayloads) {
+        relationshipPayloads.unload(modelName, id, relationshipName);
+      }
+    });
+  }
+
+  /**
+    Find the RelationshipPayloads object for the given relationship.  The same
+    RelationshipPayloads object is returned for either side of a relationship.
+
+    @example
+
+      const User = DS.Model.extend({
+        hobbies: DS.hasMany('hobby')
+      });
+
+      const Hobby = DS.Model.extend({
+        user: DS.belongsTo('user')
+      });
+
+      relationshipPayloads.get('user', 'hobbies') === relationshipPayloads.get('hobby', 'user');
+
+    @private
+  */
+  _getRelationshipPayloads(modelName, relationshipName) {
+    // TODO: lightschema this
+    let modelClass = this._store._modelFor(modelName);
```

Comment:
It's strictly internal, so by the time we get here `modelName` should already be normalized.  I'm fine with normalizing here (ie calling `this._store.modelFor` instead of `this._store._modelFor`) for simplicity's sake.